pyomo/contrib/cp/sequence_var.py

```diff
@@ -0,0 +1,151 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright (c) 2008-2024
+#  National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+import logging
+
+from pyomo.common.log import is_debug_set
+from pyomo.common.modeling import NOTSET
+from pyomo.contrib.cp import IntervalVar
+from pyomo.core import ModelComponentFactory
+from pyomo.core.base.component import ActiveComponentData
+from pyomo.core.base.global_set import UnindexedComponent_index
+from pyomo.core.base.indexed_component import ActiveIndexedComponent
+from pyomo.core.base.initializer import Initializer
+
+import sys
+from weakref import ref as weakref_ref
+
+logger = logging.getLogger(__name__)
+
+
+class SequenceVarData(ActiveComponentData):
+    """This class defines the abstract interface for a single sequence variable."""
+
+    __slots__ = ('interval_vars',)
+
+    def __init__(self, component=None):
+        # in-lining ActiveComponentData and ComponentData constructors, as is
+        # traditional:
+        self._component = weakref_ref(component) if (component is not None) else None
+        self._index = NOTSET
+        self._active = True
+
+        # This thing is really just an ordered set of interval vars that we can
+        # write constraints over.
+        self.interval_vars = []
+
+    def set_value(self, expr):
+        # We'll demand expr be a list for now--it needs to be ordered so this
+        # doesn't seem like too much to ask
```

Comment:
This is technically not requiring a `list` (which is good).  But, should there be a check for passing in a non-deterministic Sequence (e.g., `set`)?